Watchdog: refactor to Watchdog/VirtualWatchdog objects

0xc0170 opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

Follows the proposal in #10857. I don't think we need BaseWatchdog or WatchdogManager - I kept it simple, it can be easily refactored if we agree they should be there.

I havent yet touched anything else besides drivers+ platform files to get this view as small as possible (will refactor the rest of files once we get this one approved) - tests should be updated.

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

@fkjagodzinski @donatieng 

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

0xc0170 committed:
Watchdog: refactor driver

Watchdog is hardware driver. It interacts with HAL - provides wrapper to interact with the peripheral.

Provides basic functionality: start/stop, get timeout/max timeout.
It is automatically kicked by a timer, according to the timeout set in ctor.

0xc0170 committed:
VirtualWatchdog: software watchdog

Refactor old Watchdog (it was not a driver) to become VirtualWatchdog.
This is software virtual watchdog. This it the primary used watchdog in user application.

VirtualWatchdog: has-a watchdog. Initializes hw watchdog - start it when first used, stops it when there is no more VirtualWatchdog in the system -
list is empty.

Adding also check to watchdog to make sure there is only one in the system - runtime error if multiple objects created to already
running hw watchdog.

0xc0170 committed:
Remove watchdog manager

Use VirtualWatchdog instead

0xc0170 committed:
mbed: add Watchdog and VirtualWatchdog

0xc0170 committed:
Watchdog: use get_instance to access watchdog hw

0xc0170 committed:
VirtualWatchdog: fix protection macro

donatieng commented on drivers/Watchdog.h:
```diff
@@ -20,98 +20,106 @@
+    Watchdog();
+    ~Watchdog();
+
+    static void kick();
```
If the whole `Watchdog` is built using a singleton pattern (`get_instance()`), you won't need to make this static or any of the other variables below.

0xc0170 replied:
fixed

donatieng commented on drivers/Watchdog.cpp:
```diff
@@ -17,92 +17,95 @@
+    hal_watchdog_kick();
+    // VirtualWatchdog will access the watchdog process method to verify
+    // all registered users/threads in alive state */
+    VirtualWatchdog::process(((elapsed_ms <= 0) ? 1 : elapsed_ms));
```
This can be moved to the `VirtualWatchdog` itself if it can `kick()` this directly from a ticker instantiated there.

0xc0170 replied:
I probably understand the point now, checking how to decouple

donatieng reviewed (changes requested):
Thanks a lot @0xc0170 for the changes.

This looks good but I'd still be keen to decouple `Watchdog`/`VirtualWatchdog` a bit more. At the moment a user cannot build a simple app that only uses `Watchdog` and kicks it directly.

Moving some business logic from `Watchdog` to static `VirtualWatchdog` methods (or a `WatchdogManager` singleton) would let you achieve this.

0xc0170 committed:
Watchdog: cleanup after singleton usage

0xc0170 committed:
Watchdog: kick method should not have any external dep

fkjagodzinski commented on drivers/Watchdog.cpp:
```diff
@@ -17,94 +17,91 @@
-    this->_next = _first;
-    _first = this;
-    _is_initialized =  true;
 }
```
We also need a dtor definition.

0xc0170 replied:
where? to do what?

0xc0170 replied:
fixed

0xc0170 committed:
Watchdog: fix dtor

0xc0170 committed:
watchdog: add WatchdogManager

Watchdog should be usable on it's own - kicking it via ticker. No Virtual involved.
VirtualWatchdog as well - services should use this one to have multiple one in the system.

There's WatchdogManager - basically internal class to manage ticker for Watchdog.

fkjagodzinski commented on drivers/WatchdogManager.cpp:
```diff
@@ -0,0 +1,46 @@
+#if DEVICE_LPTICKER
+    /** Create singleton instance of LowPowerTicker for watchdog periodic call back of kick.
+     */
+    SingletonPtr<LowPowerTicker> _ticker;
```
```suggestion
    SingletonPtr<LowPowerTicker> WatchdogManager::_ticker;
```

0xc0170 replied:
should be fixed now

fkjagodzinski commented on drivers/WatchdogManager.cpp:
```diff
@@ -0,0 +1,46 @@
+#else
+    /** Create singleton instance of Ticker for watchdog periodic call back of kick.
+     */
+    SingletonPtr<Ticker> _ticker;
```
```suggestion
    SingletonPtr<Ticker> WatchdogManager::_ticker;
```

0xc0170 replied:
should be fixed now

fkjagodzinski reviewed (changes requested):
Minor fixes to make it compile.

0xc0170 committed:
watchdog: remove manager

Watchdog can handle callbacks - VirtualManager can attach to the tick. This should simplify the logic.

Watchdog can tick on its own using tickers.
VirtualWatchdog uses attach to get a callback when Watchdog ticks - to process own
linked list of virtual watchdogs.

0xc0170 committed:
watchdog: fix astyle errors

fkjagodzinski committed:
Watchdog & VirtualWatchdog process() update

Add uint32 argument - pass how much time already passed to the callback.

0xc0170 committed:
Watchdog: clean-up start

0xc0170 committed:
VirtualWatchdog: make sure we are the only owner of watchdog

If virtual one is not the only owner we error - init error code in driver.

0xc0170 committed:
Watchdog: fix const static

Fixing error "non-static data member initializers only available with -std=c++11 or -std=gnu++11"

0xc0170 committed:
VirtualWatchdog: fix error code and timeout

0xc0170 committed:
error: add watchdog driver error

0xc0170 committed:
Watchdog: remove dead reference file

0xc0170 committed:
Watchdog: add missing docs doxygen

0xc0170 committed:
Watchdog: refactor timeout handling

Enable kicking directly by kick() and handle timeout by using private handler.

0xc0170 committed:
VirtualWatchdog: fix doxygen docs errors

0xc0170 committed:
Watchdog: fix callback - uses ms

0xc0170 committed:
Watchdog: fix timeout - use full range for watchdog

0xc0170 committed:
VirtualWatchdog: add config option to ctor to clarify the hw timeout

0xc0170 committed:
VirtualWatchdog: fix is hw init definition

0xc0170 committed:
VirtualWatchdog: fix name for member timeout

0xc0170 committed:
VirtualWatchdog: fix default timeout value (1ms is unreal)

0xc0170 committed:
Watchdog: fix include paths - use module name

0xc0170 committed:
Watchdog: keep it simple - providing simple driver functionality

No callbacks or tickers. User is handling kick() method directly based on timeout set.
VirtualWatchdog manages ticker and callback.

donatieng commented on drivers/Watchdog.h:
```diff
@@ -20,98 +20,102 @@
+ * Watchdog &watchdog = Watchdog::get_instance();
  * watchdog.start();
  *
  * while (true) {
```
Just need `kick()`-ing that dog here :)

0xc0170 replied:
fixed !

donatieng reviewed (approved):
Looks good @0xc0170, thanks a lot for the work!

donatieng commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,134 @@
+     */
+    void kick();
+
+protected :
```
```suggestion
protected:
```

0xc0170 replied:
fixed

0xc0170 committed:
Watchdog: update unittests

0xc0170 committed:
unittest: rename watchdog to VirtualWatchdog

0xc0170 committed:
unittest: move watchdog files to drivers

Test VirtualWatchdog and Watchdog

0xc0170 committed:
unittest: remove platform watchdog (moved to drivers)

0xc0170 committed:
watchdog unittest: fix naming after refactor

0xc0170 committed:
unittest: add virtualwatchdog stub

0xc0170 committed:
VirtualWatchdog: fix missing kick and init check in process

Also syntax error for ticker definitions.

0xc0170 committed:
virtualWatchdog: fix alignment for tickers and one space

0xc0170 committed:
virtualWatchdog: fix alignment for tickers

0xc0170 committed:
Watchdog: example code - add kick()

fkjagodzinski committed:
Fix unittest errors

fkjagodzinski committed:
Fix more unittest errors

0xc0170 committed:
Watchdog: fix space in code example

fkjagodzinski committed:
Test: Watchdog: Add new driver tests

fkjagodzinski committed:
Watchdog: Fix timeout constraints

Update the start method to check 0 < timeout <= max_timeout.

fkjagodzinski committed:
Test: WatchdogManager: Remove tests

This component no longer exists in Mbed OS.

fkjagodzinski committed:
Test: Watchdog: Fix astyle errors

0xc0170 committed:
VirtualWatchdog: process fail test expects 0 as reset - no ticker running

0xc0170 committed:
Watchdog: add docs update about intention one vs multiple Watchdogs

VirtualWatchdog - multiple timers in the system - most used watchdog
Watchdog - simple driver for one system timer in the app - bootloader or similar

fkjagodzinski committed:
Test: Watchdog: Fix test_restart case

Watchdog::stop() is called in this test. If the target does not support
stopping the watchdog this case needs to be skipped.
Also improve handling ignored test cases.

bulislaw commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,142 @@
+    ~VirtualWatchdog();
+public:
+
+    /** Start an independent watchdog timer with specified parameters.
```
There's not parameters in the start function. Should it be `Start an independent watchdog timer with parameters specified in the constructor`?

0xc0170 replied:
will fix

bulislaw commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,142 @@
+    /** Constructor configured with timeout and name for this software watchdog instance.
+     *
+     * We start hardware watchdog on  the first run. This set-up uses default watchdog timeout set via config value
+     * MBED_CONF_TARGET_WATCHDOG_TIMEOUT.
```
Why do we need that config option? I mean it can be easily in constructor or start function and I'm a bit afraid of explosion of config parameters in Mbed.

0xc0170 replied:
It can be set only once - we start watchdog instance in ctor in VirtualWatchdog. 
We could expose it to ctor here but that parameter is used only once when watchdog is started. 

I could add this to ctor, as another parameter, use default value and users can use it then (I was not inclined to it due to "use it once").

0xc0170 replied:
I'll extend this, a sec

bulislaw commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,142 @@
+     *
+     * Assert without calling start.
+     */
+    void stop();
```
Should we return bool?

0xc0170 replied:
stop asserts if not initialized - the only error that can happen (catched via asserts in start/stop duo). When we would return false?

bulislaw replied:
When you can't stop the watchdog:
start -> starts the HW
start -> adds to the list
stop -> ok, removes from the list
stop -> fail as this particular HW doesn't support stopping

bulislaw replied:
Ah I see we never actually stop the HW timer.

bulislaw commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,142 @@
+
+/** \addtogroup drivers */
+
+/** VirtualWatchdog should be used for applications where you use multiple services requiring watchdog functionality.
```
We should probably mention that using both Watchdog and VirtualWatchdog at the same time is not supported.

0xc0170 replied:
will add

bulislaw commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,142 @@
+ *    // Application code
+ * }
+ * @endcode
+ * @ingroup platform
```
platform?

0xc0170 replied:
will fix

0xc0170 committed:
VirtualWatchdog: fix doxygen ingroup and ctor

0xc0170 committed:
VirtualWatchdog: add comment about exclusive access to Watchdog

0xc0170 committed:
VirtualWatchdog: provide hw timeout via ctor

bulislaw commented on drivers/VirtualWatchdog.h:
```diff
@@ -0,0 +1,146 @@
+     *  @param watchdog_hw_timeout The watchdog driver timeout - can be set only once (as soon as watchdog runs, cannot be reconfigured)
+     *
+     */
+    VirtualWatchdog(uint32_t timeout = 1000, const char *const str = NULL, uint32_t watchdog_hw_timeout = Watchdog::watchdog_timeout);
```
Why would user care about the HW timeout when using VirtualWatchdog? We should handle the complexity ourselves. I still think we shouldn't have a config option to set default timeout, for Watchdog it should be just a default parameter, but for VirtualWatchdog we should figure out the right value.

0xc0170 replied:
+1 for config value removal, that one is done , will be pushed. 

Updated, timeout is set on the first run based on timeout value.

bulislaw commented on targets/targets.json:
```diff
@@ -48,10 +48,9 @@
-                "help": "Define the timeout in ms value LowPowerTicker to do HW kick",
-                "value": "800",
-                "macro_name": "HW_WATCHDOG_TIMEOUT"
+            "watchdog-timeout": {
```
Can we remove it now?

0xc0170 replied:
removing

0xc0170 committed:
Watchdog: remove config value for timeout, set it dynamically

Use start method to set timeout

0xc0170 committed:
VirtualWatchdog: use timeout in ctor to set hw timeout

No config, neither additional parameter - the first instance of the class initalizes the peripheral.

fkjagodzinski committed:
Watchdog: Fix unittests

Compare get_timeout against the last timeout successfully set.
Merge most of the test cases into one. Do not rely on the tests
execution order.
Fix the watchdog_api_stub to keep timeout value set by init function.

0xc0170 committed:
Watchdog: remove VirtualWatchdog

We will provide documentation how to create your own VirtualWatchdog. It's simple,
create timeout and watchdog objects.

This class brought lot of discussion and questions. After our refactor we made this class
just a linked list of objects - something tickers can do as well (they already have linked list) and handling hw should be
done via Watchdog due to the limitations (timeout can be set only once per app!).

0xc0170 committed:
Watchdog: start without arg uses max timeout from a target

bulislaw reviewed (approved):
The API and structure looks good. Can someone review the tests and cpp please.

donatieng reviewed (changes requested):
@0xc0170 can you hold on please - I am not comfortable with removal of `VirtualWatchdog`

donatieng reviewed (approved):
As discussed offline, let's get this in and add extra functionality later once we got some feedback from users :)

fkjagodzinski commented on platform/mbed_error.h:
```diff
@@ -288,6 +288,7 @@ typedef enum _mbed_module_type {
     MBED_MODULE_DRIVER_PWM,
     MBED_MODULE_DRIVER_QSPI,
     MBED_MODULE_DRIVER_USB,
+    MBED_MODULE_DRIVER_WATCHDOG,
```
Seems this line is no longer needed.

0xc0170 replied:
ok once this goes into feature branch, lets remove it from there +1

fkjagodzinski reviewed (approved):
LGTM (just one minor remark, which can be addressed outside of this PR)
